src/bundle/Controller/ContentOnTheFlyController.php

```diff
@@ -114,114 +170,103 @@ public function createContentAction(
 
-    public function createUserAction(
+    /**
+     * @return \eZ\Publish\Core\MVC\Symfony\View\BaseView|\Symfony\Component\HttpFoundation\Response|null
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function editContentAction(
         Request $request,
         string $languageCode,
-        ContentType $contentType,
-        Location $parentLocation
+        int $contentId,
+        int $versionNo,
+        int $locationId
     ) {
+        $content = $this->contentService->loadContent($contentId, [$languageCode], $versionNo);
+        $versionInfo = $content->getVersionInfo();
+        $location = $this->locationService->loadLocation($locationId);
+
+        $contentType = $this->contentTypeService->loadContentType(
+            $content->contentInfo->contentTypeId,
+            $this->userLanguagePreferenceProvider->getPreferredLanguages()
+        );
+
+        if ((new ContentTypeIsUser($this->userContentTypeIdentifiers))->isSatisfiedBy($contentType)) {
+            return $this->forward('EzPlatformAdminUiBundle:UserOnTheFly:editUser', [
+                'languageCode' => $languageCode,
+                'contentId' => $contentId,
+                'versionNo' => $versionNo,
+                'locationId' => $locationId,
+            ]);
+        }
+
         $language = $this->languageService->loadLanguage($languageCode);
-        $parentGroup = $this->userService->loadUserGroup($parentLocation->contentId);
 
-        $data = (new UserCreateMapper())->mapToFormData($contentType, [$parentGroup], [
-            'mainLanguageCode' => $language->languageCode,
+        $contentUpdate = (new ContentUpdateMapper())->mapToFormData($content, [
+            'languageCode' => $languageCode,
+            'contentType' => $contentType,
         ]);
-        $form = $this->createForm(UserCreateType::class, $data, [
-            'languageCode' => $language->languageCode,
-            'mainLanguageCode' => $language->languageCode,
-         ]);
+
+        $form = $this->createForm(
+            ContentEditType::class,
+            $contentUpdate,
+            [
+                'languageCode' => $languageCode,
+                'mainLanguageCode' => $content->contentInfo->mainLanguageCode,
+                'drafts_enabled' => true,
+            ]
+        );
 
         $form->handleRequest($request);
 
-        if ($form->isSubmitted() && $form->isValid() && $form->getClickedButton()) {
-            $this->userActionDispatcher->dispatchFormAction($form, $data, $form->getClickedButton()->getName());
-            if ($response = $this->userActionDispatcher->getResponse()) {
-                return $response;
-            }
+        if (!$versionInfo->isDraft()) {
+            throw new InvalidArgumentException('Version', 'The status is not draft');
```

Comment:
should it not be `BadStateException`, also in _classic_ flow we create draft in that case, how is this different?

Reply:
in admin-ui, there is flow like this
```php
                if (!$versionInfo->isDraft()) {
                    $contentDraft = $this->contentService->createContentDraft($contentInfo, $versionInfo, null, $language);
}
```
So for some reason, we are handling cases when provided `versionInfo` is not a draft.

Reply:
Ping @webhdx . Any reason why it is like that for Edit on the Fly?